Multi stage docker build

jrajahalme opened the pull request:
Use staged Docker builds to use prebuilt Envoy dependencies.

Staged Docker build requires the build to be done on new enough Docker
(e.g. Docker 17.09), but k8s does not support any Docker version that
is new enough. As a workaround the docker build step in k8s testsuite
is performed in a new builder VM using newer docker. The k8s VM keep
using older docker but pull Cilium images from the new builder VM. The
builder VM can be killed after k8s VMs have pulled the images so the
overall resource consumption should not increase significantly.

As part of this change we give the main Cilium directory to docker
build, but with a generated .dockerignore file that avoids sending
over the git directory.

The new Dockerfile refers to prebuilt builder and runtime images,
built with Dockerfile.builder and Dockerfile.runtime, respectively
(both located in contrib/packaging/docker/). The prebuilt images are
to be built using "make docker-image-builder" and "make
docker-image-runtime" at the repo top level directory. Newly built
builder and runtime images need to be pushed to Docker Hub before they
can be used by the main Dockerfile for building Cilium using the date
(format YYYY-MM-DD) as the image tag. Finally, the main Dockerfile
needs to be updated in the repo with the new tags.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>


```release-note
Use multi - staged Docker builds to use prebuilt Envoy dependencies.
```

aanm commented on tests/k8s/Vagrantfile:
```diff
@@ -79,6 +87,28 @@ Vagrant.configure(2) do |config|
+            # Do not inherit DNS server from host, use proxy
+            vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
+            vb.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
+            vb.memory = 3072
```
Do we need this amount of memory and cpu for the registry?

jrajahalme replied:
It is not just registry, but also the build VM. On my testing C++ build of Cilium Envoy filters fails with 2GB, but succeeds with 3GB. Also, we delete this VM before we we start running the tests in the other two, so the peak memory use should not increase much.

ianvernon replied:
An enhancement as a follow-up could be to use a lighter-weight image - doesn't have to be a part of this ticket, though IMO.

jrajahalme replied:
renamed the new VM as Ray suggested. It is now called 'builder' which should avoid future confusion here as well :-)

jrajahalme replied:
I tested builder with 1 CPU and 1GB of memory and it works, so I'll modify this PR accordingly. I'll also roll back the change in number of CPUs on the k8s nodes, I wanted to test if it impacts the runtime but then did not follow through actually timing it.

raybejjani commented on Dockerfile:
```diff
@@ -1,89 +1,43 @@
+# versions to be built while allowing the new versions to make changes
+# that are not backwards compatible.
+#
+FROM cilium/cilium-builder:2017-12-14 as builder
```
I'd suggest the tag also include a revision for that day. This allows us to push hotfixes to images more than once a day, and then avoid ambiguity over which actual image is being run on machines during these bugfix pushes. An alternative would be to have hour:minute on there too but that has a similar problem on a minute time scale.

jrajahalme replied:
Builder and runtime base images are not very likely to be re-created very often. Especially for the builder image I do not see a scenario where hotfixes would be needed, as bazel build cache will just become less useful it becomes out-of-date.

Generally, using the git sha as the tag would be most accurate, but even that does not cover the fact that some runtime and build tool dependencies are not explicitly managed, but we install the latest version available via apt.

So I though using a date would strike a good balance here, implicitly conveying a notion of freshness. Nothing prevents adding retagging the images manually after a build with a more detailed tag either.

That said, the above is all there is to the choice of the automatically generated tag, open to more discussion and viewpoints as always :-)

raybejjani replied:
This will totally work as-is. I was projecting some mostly forgotten issue I once had (all I do remember is having to retrofit the revision onto the date tag). In this case, especially, it's pretty easy to fix it with the manual tagging, as you said. +1

raybejjani commented on Makefile:
```diff
@@ -1,13 +1,15 @@
 GOLANGVERSION = $(shell go version 2>/dev/null | grep -Eo '(go[0-9].[0-9])')
-GOLANG_SRCFILES=$(shell for pkg in $GOFILES; do find $(pkg) -name *.go -print; done | grep -v /vendor/)
-BPF_SRCFILES=$(shell find bpf/ -name *.[ch] -print)
+GOLANG_SRCFILES=$(shell for pkg in $(subst github.com/cilium/cilium/,,$(GOFILES)); do find $$pkg -name *.go -print; done | grep -v /vendor/)
```
Probably out of scope for this PR but this might be simpler with `go list -f '{{ join .GoFiles " "}}' ./...`

jrajahalme replied:
I tried this and it produces a list without any directory hierarchy: ```
can't load package: package github.com/cilium/cilium/common/tests: C source files not allowed when not using cgo or SWIG: bpf-event-test.c
cilium_client.go
daemon_client.go get_config_parameters.go get_config_responses.go get_debuginfo_parameters.go get_debuginfo_responses.go get_healthz_parameters.go get_healthz_responses.go patch_config_parameters.go patch_config_responses.go
...
```

raybejjani replied:
boo. Ignore me. I thought I'd tested it and it had fuller paths but I think there's some other template combination we need to throw in. Not worth it to investigate further.

raybejjani commented on Makefile:
```diff
@@ -1,13 +1,15 @@
-GOLANG_SRCFILES=$(shell for pkg in $GOFILES; do find $(pkg) -name *.go -print; done | grep -v /vendor/)
-BPF_SRCFILES=$(shell find bpf/ -name *.[ch] -print)
+GOLANG_SRCFILES=$(shell for pkg in $(subst github.com/cilium/cilium/,,$(GOFILES)); do find $$pkg -name *.go -print; done | grep -v /vendor/)
+BPF_SRCFILES=$(shell find bpf -name *.[ch] -print)
```
Probably out of scope for this PR but this might be more reliable using BPF_FILES from https://github.com/cilium/cilium/blob/master/daemon/bpf.sha#L3

jrajahalme replied:
Done!

raybejjani commented on Makefile:
```diff
@@ -110,30 +112,41 @@ tags: $(GOLANG_SRCFILES) $(BPF_SRCFILES)
+
+docker-image: clean GIT_VERSION envoy/SOURCE_VERSION
+	grep -v -E "(SOURCE|GIT)_VERSION" .gitignore >.dockerignore
+	echo ".*" >>.dockerignore # .git prunend out
```
typo in prunend

jrajahalme replied:
Fixed, thanks!

raybejjani commented on tests/k8s/Vagrantfile:
```diff
@@ -79,6 +87,28 @@ Vagrant.configure(2) do |config|
+    # builds and not running k8s.  This VM will be idle after k8s[12] have pulled the
+    # newly built cilium image.
+    config.vm.define "k8s0#{$build_id_name}" do |s|
+        s.vm.hostname = "k8s-0"
```
Why not give this a more descriptive name? It's transient, as you say, and it doesn't run k8s. envoy-builder-registry (or something) would make it clearer if, for example, we leak VMs on a build machine one day.

jrajahalme replied:
Most likely I initially thought I could use the same loop that loops from 1 to 2 to also configure the new VM. That turned out not to be the case, so I'll experiment with a more descriptive name.

Either way, the certs need to be regenerated as the IP address of the registry node needs to be in a cert. I guess we could use a separate, completely new cert for the registry node and leave the k8s cluster cert untouched, but that would have been more work and more new files in the repo. I'm leaning for someone more experienced working with certs fixing this up later, though.

raybejjani replied:
Ah, I didn't realise there was this much interrrelation between the certs and nodes. I see you renamed it below which more than addresses my primary concern. Thanks.

raybejjani reviewed (approved):
Nothing blocking.
I'm curious, why did all the test CA certs change?

ianvernon reviewed (approved):
Amazing work!

aanm commented on contrib/vagrant/start.sh:
```diff
@@ -596,6 +596,9 @@ create_k8s_config
 cd "${dir}/../.."
 
+# Store git version in a file so that .git does not need to be synced to the VM
+make GIT_VERSION envoy/SOURCE_VERSION
```
What happens when using NFS and not rsync?

jrajahalme replied:
GIT_VERSION file is used only if .git directory does not exist. However, I see that envoy/SOURCE_VERSION is used by the bazel scripts whenever it exists, so I'll have to hix that!

jrajahalme replied:
I modified the envoy/bazel/get_workspace_status script to ignore envoy/SOURCE_VERSION if git repo is available.

jrajahalme committed:
tests/k8s: Run docker build in a separate VM

Create a new VM 'k8s0' to run docker registry and docker build with
new enough Docker CE that supports staged builds.  'k8s1' and 'k8s2'
are still running older Docker as required for Kubernetes.  'k8s0' is
destroyed after the other nodes have pulled the Cilium image to limit
resource use during k8s tests.

The number of available CPUs for 'k8s1' and 'k8s2' is increased to 4
each in hopes of making the tests run a little faster.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
Dockerfile: Incremental build.

Use staged Docker builds to use prebuilt Envoy dependencies.

Staged Docker build requires the build to be done on new enough Docker
(e.g. Docker 17.09), but k8s does not support any Docker version that
is new enough. As a workaround the docker build step in k8s testsuite
is performed in a new builder VM using newer docker. The k8s VM keep
using older docker but pull Cilium images from the new builder VM. The
builder VM can be killed after k8s VMs have pulled the images so the
overall resource consumption should not increase significantly.

As part of this change we give the main Cilium directory to docker
build, but with a generated .dockerignore file that avoids sending
over the git directory.

The new Dockerfile refers to prebuilt builder and runtime images,
built with Dockerfile.builder and Dockerfile.runtime, respectively
(both located in contrib/packaging/docker/). The prebuilt images are
to be built using "make docker-image-builder" and "make
docker-image-runtime" at the repo top level directory. Newly built
builder and runtime images need to be pushed to Docker Hub before they
can be used by the main Dockerfile for building Cilium using the date
(format YYYY-MM-DD) as the image tag. Finally, the main Dockerfile
needs to be updated in the repo with the new tags.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
Makefile: Fix typo.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
Dockerfile: Enable Envoy by default.

This makes the release and k8s tests use Envoy by
default. Runtime-tests are still run also with the go proxy code.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
Vagrantfile: Avoid syncing .git for CI builds.

At least locally rsyncing the cilium directory takes a long time. Try
to cut this down by not syncinfg the .git directory when BUILD_NUMBER
is defined.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
tests/k8s: Rename 'k8s0' as 'builder'

'k8s0' VM is running k8s, but as the docker builder and registry.
Rename it as 'builder' to reflect this.

Suggested-by: Ray Bejjani <ray@covalent.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
Makefile: Use daemon/bpf.sha.

Use daemon/bpf.sha as the authoritative list of bpf files.

Suggested-by: Ray Bejjani <ray@covalent.io>

jrajahalme committed:
envoy: Use git version when git is available.

Use SOURCE_VERSION only if git is not available. For actual build the
SHA of the git HEAD is always used, but for dependencies build the SHA
of the Envoy dependency is used instead.

GIT_VERSION was already ignored if .git directory exists.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
tests/k8s/Vagrantfile: Roll back VM CPUs, memory.

Revert changes for k8s VM CPU counts and make the builder VM smaller.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

manalibhutiyani commented on Dockerfile:
```diff
@@ -1,89 +1,43 @@
+WORKDIR /go/src/github.com/cilium/cilium
+ADD . ./
+#
+# Please do not add any dependency updates before the 'make install' here,
```
If this is really important and will tend to break envoy, do you want add a comment which will get ones attention
```
# ********** WARNING ********** 
# Please do not add any dependency updates before the 'make install' here,
```

Feel free to ignore, if you feel this is unnecessary

manalibhutiyani reviewed (approved):
Could you please rebase the branch?
